### PR TITLE
API of GAN

### PR DESCRIPTION
**Pull request by zchen0211**
Based on some popular models in TensorFlow and Torch, we design the logic of GAN API.

It is too complex to realize in an op, so I use python API.

### CODE REVIEW

File: doc/design/gan_api.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
+'''
+GAN implementation, just a demo.
+'''
+# pd for short, should be more concise.
+from paddle.v2 as pd
```
Comment (Superjom) [inline doc/design/gan_api.md]: use
\```python
\```

that will give a python syntax highlight.

File: doc/design/gan_api.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    for epoch in xrange(10000):
+      for batch_id in range(N / batch_size):
+        idx = ...
+        # sample a batch
+        batch_im, batch_label = data_X[idx:idx+batch_size], data_y[idx:idx+batch_size]
+        # sample z
+        batch_z = np.random.uniform(-1., 1., [batch_size, z_dim])
+
+        if batch_id % 2 == 0:
+          sess.run(d_optim,
```
Comment (Superjom) [inline doc/design/gan_api.md]: use our interface, `pd.eval`?

File: doc/design/gan_api.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+# pd for short, should be more concise.
+from paddle.v2 as pd
+import numpy as np
+import logging
+
+X = pd.data(pd.float_vector(784))
+
+# Conditional-GAN should be a class. 
+### Class member function: the initializer.
+class DCGAN(object):
```
Comment (Superjom) [inline doc/design/gan_api.md]: no need to be a class, simply 3 functions should be more simple.

File: doc/design/gan_api.md (hunk 2 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: Done.

File: doc/design/gan_api.md (hunk 1 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: Done.

File: doc/design/gan_api.md (hunk 3 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: We may have a discussion later. I refer to several GAN models on github, pretty much all looks like this.

File: doc/design/gan_api.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+### Class member function: Generator Net
+```python
+def generator(self, z, y = None):
+
+    # Generator Net
+    if not self.y_dim:
+      z = pd.concat(1, [z, y])
+      
+    G_h0 = pd.fc(z, self.G_w0, self.G_b0)
```
Comment (Superjom) [inline doc/design/gan_api.md]: fc layer support a `param_name` argument and generate a unique name , just use it like
```python
G_h0 = pd.fc(z, size=y_dim, param_name="G_fc0")
```

call `generator` twice will not generate different parameters.

and each `fc` needn't declare parameters somewhere, so is `batchnorm`.

File: doc/design/gan_api.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
+# Design for GAN
+
+GAN (General Adversarial Net) is an important model for unsupervised learning and widely used in many areas.
```
Comment (jacquesqiao) [inline doc/design/gan_api.md]: maybe add a link to the official definition

File: doc/design/gan_api.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
+# Design for GAN
+
+GAN (General Adversarial Net) is an important model for unsupervised learning and widely used in many areas. 
+
+It contains several important machine learning concepts, including building and running subgraphs, dependency tracing, different optimizers in one executor and so forth.
+
+In our GAN design, we wrap it as a user-friendly easily customized python API to design different models. We take the conditional DC-GAN as an example due to its good performance on image generation.
```
Comment (jacquesqiao) [inline doc/design/gan_api.md]: also, add a link to DC-GAN

File: doc/design/gan_api.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
+# Design for GAN
+
+GAN (General Adversarial Net) is an important model for unsupervised learning and widely used in many areas. 
+
+It contains several important machine learning concepts, including building and running subgraphs, dependency tracing, different optimizers in one executor and so forth.
```
Comment (jacquesqiao) [inline doc/design/gan_api.md]: I think `contain` is not good, maybe `It will use ...`?

File: doc/design/gan_api.md [hunk 8]
```diff
@@ -19,12 +19,12 @@ In our GAN design, we wrap it as a user-friendly easily customized python API to
 | softmax loss (done)       | ?                 | Y        |
 | reshape op (done)         | ?                 | Y        |
```
Comment (Superjom) [inline doc/design/gan_api.md]: I think it is better to create an issue to place this information.
use a `- [ ] zzz` to create checkboxes to follow up the process.

File: doc/design/gan_api.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+- discriminator(image):
+Given an image, decide if it is from a real source or a fake one. 
+Returns a 0/1 binary label.
+
+- build_model(self):
+build the whole GAN model, define training loss for both generator and discrimator.
+
+## Discussion on Engine Functions required to build GAN
+- Trace the ternsor and variable dependency in the engine executor. (Very critical, otherwise GAN can'be be trained correctly)
```
Comment (QiJune) [inline doc/design/gan_api.md]: ternsor --> tensor

File: doc/design/gan_api.md (hunk 5 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: Thanks. Done.

File: doc/design/gan_api.md (hunk 6 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: Done.

File: doc/design/gan_api.md (hunk 7 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: Thanks. Done.

File: doc/design/gan_api.md (hunk 9 above)
Comment (zchen0211) [reply doc/design/gan_api.md]: Done.
